### PR TITLE
Curry jit, grad, value_and_grad and vmap.

### PR DESCRIPTION
**Pull request by BuddenD**
Allows use as decorators with kwargs, e.g.:

`@jax.jit(static_argnums=(1,))
`
as opposed to the current, less pythonic:

`@functools.partial(jax.jit, static_argnums=(1,))
`
Also updates all internal usages.

### CODE REVIEW

File: jax/BUILD [hunk 1]
```diff
@@ -43,7 +43,10 @@ pytype_library(
         ],
     ),
     srcs_version = "PY3",
-    deps = ["@org_tensorflow//tensorflow/compiler/xla/python:xla_client"],
+    deps = [
+        "//third_party/py/toolz",
```
Comment (mattjj) [inline jax/BUILD:47]: It seems odd that a toolz dependence would need to be added here, whereas no other jax Python dependencies need to be. Are you sure this is needed?

**Comment by shoyer** [comment]
The alternative to adding a toolz dependency would be to write our own helper function, e.g., a decorator that produces something like:
```python
def wrapper(*args, **kwargs):
  if args or 'fun' in kwargs:
    return jit(*args, **kwargs)
  else:
    if args:
      raise ValueError("can't use positional arguments when currying")
    return partial(jit, **kwargs)
```

**Comment by mattjj** [comment]
Thanks for proposing and prototyping this! Recording my thoughts here for posterity, though we discussed more in our chat rooms. Not everyone agrees with the reasoning or conclusion here!

I don't think we should merge this. My primary concern is that there are unforeseen edge cases that will end up harming usability on net. Once we merge a new api that users start relying on, if edge cases later emerge it may be difficult or impossible to revise our decision and remove it. It's possible that there are _no edge cases at all_ to this convenience layer, but the `tools.curry` logic [looks tricky enough](https://github.com/pytoolz/toolz/blob/ea3ba0d60a33b256c8b2a7be43aff926992ffcdb/toolz/functoolz.py#L300-L335) that it's not easy for me to make that claim without deeper investigation. To be confident, I'd want to thoroughly understand the `toolz.curry` logic, and still probably brainstorm and write tests for all the evil edge cases I can think up. That would take time, and I don't think the potential upside of this convenience layer justifies that kind of time investment given our other priorities.

A distant secondary concern is performance. `jit` dispatch time isn't good now, but because it's relatively simple we know how to make it fast, and we've already started working on that. I don't want to add another wrapper from an external library to that path now, since it's another potential blocker to fast dispatch.

The current solution to the issue at hand here, with `@partial(jit, ...)`, is simple and predictable. It may be unfamiliar at first, but it's easy to learn and once learned it's easy to understand and predict its behavior. Moreover, any user who wants a curried version can get it as simple as:

```python
import toolz
from jax import jit

jit = tools.curry(jit)
```

That is, we don't need to bundle these things together for users. They can use composition instead.

While I don't think we should merge this, I appreciate the time you put into making a concrete proposal we could debate. Without it, we wouldn't have been able to weigh alternatives! Investigating this carefully made JAX better.
